scheduler/src/cook/mesos/util.clj

```diff
@@ -24,7 +24,59 @@
             [metatransaction.core :refer (db)]
             [metrics.timers :as timers]
             [plumbing.core :as pc :refer (map-vals map-keys)])
-  (:import [java.util Date]))
+    (:import
+      [com.google.common.cache Cache]
+      [com.google.common.cache CacheBuilder]
+      [java.util.concurrent TimeUnit]
+      [java.util Date]))
+
+
+(defn new-cache []
+  "Build a new cache"
+  (-> (CacheBuilder/newBuilder)
+      (.maximumSize 1000000)
+      ;; if its not been accessed in 2 hours, whatever is going on, its not being visted by the
+      ;; scheduler loop anymore. E.g., its probably failed/done and won't be needed. So,
+      ;; lets kick it out to keep cache small.
+      (.expireAfterAccess 2 TimeUnit/HOURS)
+      (.build)))
+
+(defn lookup-cache!
+      "Generic cache. Caches under a key (extracted from the item with extract-key-fn. Uses miss-fn to fill
+      any misses. Caches only postive hits where both functions return non-nil"
+      [cache-atom extract-key-fn miss-fn item]
+      (let [key (extract-key-fn item)
+            ^Cache cache @cache-atom]
+           (if key
+             (if-let [result (.getIfPresent cache key)]
+                     result ; we got a hit.
+                     (let [new-result (miss-fn item)]
+                          ; Only cache non-nil
+                          (when new-result
+                                (.put cache key new-result))
+                          new-result))
+             (miss-fn item))))
+
+(defn lookup-cache-datomic-entity!
+      "Specialized function for caching where datomic entities are the key. Extracts :db/id so that
+      we don't keep the entity alive in the cache"
+      [cache miss-fn entity]
+      (lookup-cache! cache :db/id miss-fn entity))
+
+
+; Fake. Are actually initialized in flush-caches
+(defonce job-ent->resources-cache (atom (new-cache)))
```

Comment:
Should these be dynamic vars instead of atoms?  We could use `binding` in the unit tests, which is scoped around the code that need independent caches, as opposed to remembering to flush before a test because "we know" someone polluted the cache already.

Reply:
No. We want these to be shared across threads. E.g., as a side effect, these caches make offer processing and the metrics dashboard faster/cheaper.